PAWS/ECGRecordings/ECGModule.swift

```diff
@@ -96,4 +96,133 @@ class ECGModule: Module, DefaultInitializable, EnvironmentAccessible {
+                    symptoms: symptoms,
+                    voltageMeasurements: voltageMeasurements
+                )
+            )
+        } else {
+            resource = try sample.resource
+        }
+        
+        if let mockWebService {
+            let encoder = JSONEncoder()
+            encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
+            let jsonRepresentation = (try? String(data: encoder.encode(resource), encoding: .utf8)) ?? ""
+            try await mockWebService.upload(path: "healthkit/\(sample.uuid.uuidString)", body: jsonRepresentation)
+        } else {
+            try await standard.healthKitDocument(id: sample.id).setData(from: resource)
+        }
+    }
+    
+    func upload(electrocardiogram: HKElectrocardiogram) async {
+        var supplementalMetrics: [HKSample] = []
+        
+        do {
+            try await upload(sample: electrocardiogram)
+            
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingPulseRates) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingPhysicalEffort) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingStepCount) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingActiveEnergy) ?? [])
+            
+            if let precedingVo2Max = try? await electrocardiogram.precedingVo2Max {
+                supplementalMetrics.append(precedingVo2Max)
+            }
+            
+            for supplementalMetric in supplementalMetrics {
+                do {
+                    try await upload(sample: supplementalMetric)
+                } catch {
+                    logger.log("Could not upload \(supplementalMetric.sampleType): \(error)")
+                    let content = UNMutableNotificationContent()
+                    content.title = "Upload Error"
+                    content.body = "Sample could not be uploaded \(supplementalMetric.sampleType.description) (\(supplementalMetric.uuid.uuidString) at \(Date.now.formatted(date: .numeric, time: .complete)): \((supplementalMetric as? HKQuantitySample)?.quantity.description ?? "Unknown")"
+                    let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
+                    try? await UNUserNotificationCenter.current().add(request)
+                }
+            }
+        } catch {
+            logger.log("Could not access HealthKit sample: \(error)")
+            await addECGMessage(for: electrocardiogram)
+        }
+    }
+    
+    
+    /// Creates a notification with a title and body message when there is an error accessing a HealthKit sample.
+    /// - Parameter electrocardiogram: The `HKElectrocardiogram` object for which the error occurred.
+    func addECGMessage(for electrocardiogram: HKElectrocardiogram) async {
```

Comment:
This function can 
1. Be private as we will only use it in this scope
2. I would suggest to pass in the error and display it in the notification.
3. I would suggest to make the error message a bit more human-friendly. E.g. instead of the identifier we could say something like "Electrocardiogram recorded on DATE+TIME could not be uploaded. Please open up the PAWS app to re-upload the recorded electrocardiogram. \n\n ERROR_DESCRIPTION". In the long run, we can remove the error description but it might be good for debuting purposes for now.